### PR TITLE
Plymouth splash screen

### PR DESCRIPTION
**Pull request by abbradar**
###### Motivation for this change

Because we can and it looks nice. Enable with `boot.plymouth.enable = true`.
###### Things done
- [x] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
###### TODO:
- [x] Test on various configurations. Notice that it's recommended to add your video driver module to `boot.initrd.kernelModules` -- for example `i915` for Intel;
- [ ] Make GDM smoothly appear after it (there is support but I haven't tried to enable it or test);
- [x] Test with LUKS (implemented password query but not tested, YubiKey too);
- [x] Better style (I've added our logo but haven't tweaked its size, colors etc.);
- [x] Test most likely broken emergency mode -- we patch out `plymouth quit` before it in our `systemd`;
- [x] Investigate if it's possible to show systemd service failures properly;
- [x] Fix stage-1 `fail`.
###### Implementation details:

This links various resources, including binary Plymouth plugins, to `/etc`. I've done this because we want to have trimmed resources package for including in initrd -- so in it not all plugins, themes etc are available.

### CODE REVIEW

File: pkgs/os-specific/linux/plymouth/default.nix [hunk 1]
```diff
@@ -1,43 +1,54 @@
...
       --localstatedir=/var
-      --with-log-viewer
+      --with-logo=/etc/plymouth/logo.png
+      --with-release-file=/etc/os-release
       --without-system-root-install
       --without-rhgb-compat-link
       --enable-tracing
       --enable-systemd-integration
       --enable-pango
+      --enable-gdm-transition
```
Comment (matthewbauer) [inline pkgs/os-specific/linux/plymouth/default.nix]: Does GDM need to be compiled with a plymouth flag as well?

Comment (abbradar) [inline pkgs/os-specific/linux/plymouth/default.nix]: Yes, it needs to. I haven't tried Plymouth with GDM yet -- it works with SDDM for me as is, so I guess it's OK.

File: pkgs/os-specific/linux/plymouth/default.nix [hunk 2]
```diff
@@ -1,43 +1,54 @@
...
-    sed -e "s#\$(\$PKG_CONFIG --variable=systemdsystemunitdir systemd)#$out/etc/systemd/system#g" \
-      -i configure.ac
-  '';
+  buildInputs = [
+    gtk3 udev systemd
+  ];
 
   postPatch = ''
+    sed -i \
+      -e "s#\$(\$PKG_CONFIG --variable=systemdsystemunitdir systemd)#$out/etc/systemd/system#g" \
```
Comment (matthewbauer) [inline pkgs/os-specific/linux/plymouth/default.nix]: You should be able to skip this patch part using --with-systemdunitdir=$out/etc/systemd/system

see https://cgit.freedesktop.org/plymouth/tree/configure.ac#n125
and https://github.com/NixOS/nixpkgs/blob/39ebb01d6e8d19b042fe65d303234d0db27877eb/pkgs/os-specific/linux/bluez/default.nix#L30

Comment (abbradar) [inline pkgs/os-specific/linux/plymouth/default.nix]: Interesting, there's no such lines in Plymouth 0.9.2. I guess it was added recently and is not in the release yet...

**Comment by abbradar** [comment]
I performed some testing and also gave up on requesting LUKS password via Plymouth for now. The problem is YubiKey, which I cannot test properly and don't want to break people' systems.

**Comment by abbradar** [comment]
GDM doesn't work for me at all -- I guess it needs GNOME 3. Good news -- enabling Plymouth support in GDM doesn't add to its closure size. If there is someone with GNOME 3 who wants to try this out, please do!

**Comment by abbradar** [comment]
Regarding style: I should admit that I'm notoriously bad at design, so I could use some help here with colors (`--with--background-color`, `--with-background-start-color-stop`, `--with-background-end-color-stop` -- each accepting HTML-like `0xabcdef` colors) and maybe a different default logo. To test things you can run `sudo plymouthd` and then `sudo plymouth start-splash` from X11 -- it would create a GTK window. To stop -- `sudo plymouth quit`.

**Comment by abbradar** [comment]
#16788 gets systemd emergency support covered. We should wait for it to be merged first.

**Comment by abbradar** [comment]
Also, many thanks to @aske for testing Plymouth with encryption!

**Comment by abbradar** [comment]
After more fiddling and playing with settings I think I've finished this work!

* We now have an option to set console fonts and keymaps early (before LUKS passphrase entering). This is needed because Plymouth prevents doing this while it's running. Also may be a generally nice option.
* Background color is now set to black.

It appears Plymouth reacts correctly to systemd unit status updates. I can't test GDM integration but don't have enough motivation to work on it further -- maybe I'll continue after a major GDM upgrade, which should be coming soon.

**Comment by matthewbauer** [comment]
Looks good to me! Make sure to close #7400 when this gets merged.

**Comment by abbradar** [comment]
We should wait till staging hits master with needed fixes for systemd till we can merge this.

I have also splitted early consolefonts to #16801 -- turns out they are not always needed and can even make it worse (if you don't add display driver module to initrd).

**Comment by abbradar** [comment]
Merged to staging in https://github.com/NixOS/nixpkgs/commit/1848bfc92d247feff132f92b5580613eec9419d3 (I figured out why not merge there instead of waiting till needed patches hit master).

**Pull request closed without merging**